### PR TITLE
2ccvc9r collect data from cli

### PR DESCRIPTION
**Pull request by shubham-shay-te**

### CODE REVIEW

File: core_lib/helpers/shell_utils.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+def check_default(value, default_value):
+    if not value and default_value is not None:
+        return default_value
+    else:
+        return value
+
+
+def input_yes_no(title, default_value: bool) -> bool:
+    converted_default_value = 'yes' if default_value else 'no'
+    user_input = input(f'{title} (yes/no) [default: {converted_default_value}]: ')
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:13]: this should be also in the while. it's duplicate code 
(don't call your variable `ask` this is an example)
```Python
ask = True 
while ask: 
    user_input = input(...
    ask = ..... (True if ask again False if we have the value)
```

File: core_lib/helpers/shell_utils.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def check_default(value, default_value):
+    if not value and default_value is not None:
+        return default_value
+    else:
+        return value
+
+
+def input_yes_no(title, default_value: bool) -> bool:
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:11]: `default_value: bool = None` this means the user must type input

File: core_lib/helpers/shell_utils.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        if not user_input:
+            user_input = converted_default_value
+    if user_input.lower() in ['yes', 'y']:
+        return True
+    elif user_input.lower() in ['no', 'n']:
+        return False
+
+
+def input_str(title, default_value=None) -> str:
+    user_input = input(f'{title} [default: {default_value}]: ')
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:27]: if default_value is None do we need to print the word `default: `?

File: core_lib/helpers/shell_utils.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    user_input = check_default(user_input, default_value)
+    while not str(user_input):
+        user_input = input(f'{title} [default: {default_value}]: ')
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:30]: it's better to write code once. see above loop

File: core_lib/helpers/shell_utils.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        user_input = check_default(user_input, default_value)
+    return str(user_input)
+
+
+def input_int(title, default_value=None) -> int:
+    user_input = input(f'{title} [default: {default_value}]: ')
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:36]: dont print default if not provided

File: core_lib/helpers/shell_utils.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    user_input = check_default(user_input, default_value)
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:37]: no code duplication

File: core_lib_generator/core_lib_template_generator.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    CRUDSoftDelete = 4
+    CRUDSoftDeleteToken = 5
+
+
+def generate_db_template() -> dict:
+    db_template = {}
+    add_db = True
+    while add_db:
+        db_name = input_str('What is the name of the DB connection?')
+        while db_name in db_template:
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:47]: id the db name exists. tell the user it's already exists !

File: core_lib/helpers/shell_utils.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        return False
+    elif int(value) not in enum_values:
+        return False
+    else:
+        return True
+
+
+def input_list(list_value, title):
+    [print(f'{list_value.index(i) + 1}-{i}') for i in list_value]
+    user_input = input(f'From the following list, select the relevant number {title}: ')
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:66]: `From the following list, select the relevant number` this need to be provided with the title

File: core_lib/helpers/shell_utils.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+def input_int(title, default_value=None) -> int:
+    user_input = input(f'{title} [default: {default_value}]: ')
+    user_input = check_default(user_input, default_value)
+    while not is_int(user_input):
+        user_input = input(f'{title} [default: {default_value}]: ')
+        user_input = check_default(user_input, default_value)
+    return int(user_input)
+
+
+def input_enum(enum_class, title) -> int:
```
Comment (shay-te) [inline core_lib/helpers/shell_utils.py:44]: default value?

File: core_lib_generator/core_lib_template_generator.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+                    'session': {
+                        'pool_recycle': db_pool_recycle,
+                        'pool_pre_ping': db_pool_pre_ping,
+                    },
+                    'url': {
+                        'protocol': DBTypes(db_type).name.lower(),
+                    },
+                }
+                continue
+        default_db_ports = {
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:75]: define this outside the function under the DBTypes enum. no need to recreated every function call

File: core_lib_generator/core_lib_template_generator.py [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+            DBTypes.MySQL.name: 3306,
+            DBTypes.Oracle.name: 1521,
+            DBTypes.MSSQL.name: 1433,
+            DBTypes.Firebird.name: 3050,
+            DBTypes.Sybase.name: 5000,
+        }
+        db_port = input_int('Enter the port no. of your DB', default_db_ports[DBTypes(db_type).name])
+        db_host = input_str('Enter host of your DB', 'localhost')
+        db_username = input_str('Enter your DB username', 'user')
+        db_password = input_str('Enter your DB password', 'password')
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:87]: don;t offer default for password. let the user type it in

File: core_lib_generator/core_lib_template_generator.py [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+        add_db = input_yes_no('\nDo you want to add another DB connection?', False)
+        db_template[db_name.lower()] = {
+            'env': {
+                f'{db_name.upper()}_USER': db_username,
+                f'{db_name.upper()}_PASSWORD': db_password,
+                f'{db_name.upper()}_PORT': db_port,
+                f'{db_name.upper()}_DB': db_name,
+                f'{db_name.upper()}_HOST': db_host,
+            },
+            'config': {
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:98]: config needs to be with the same structure as in the real config file. 
```yaml
      log_queries: false
      create_db: false
      session:
        pool_recycle: 3600
        pool_pre_ping: false
      url:
        protocol: 'sqlite'
```

File: core_lib_generator/core_lib_template_generator.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+    return {
+        'env': {
+            'SOLR_HOST': solr_host,
+            'SOLR_PORT': solr_port,
+        },
+        'config': {
+            'protocol': solr_protocol,
+            'host': f'${{oc.env:SOLR_HOST}}',
+            'port': f'${{oc.env:SOLR_PORT}}',
+            'file': solr_file,
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:125]: solr have no file name in the connection. and it's http predefined for the protocol

File: core_lib_generator/core_lib_template_generator.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        },
+    }
+
+
+def generate_cache_template() -> dict:
+    cache_type = input_enum(CacheTypes, 'for cache type')
+
+    if cache_type == 1:
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:133]: NO MAGIC NUMBERS
CacheTypes.Memcached.value

File: core_lib_generator/core_lib_template_generator.py [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+        }
+    elif cache_type == 2:
+        print(f'\nCache type {CacheTypes(cache_type).name}')
+        return {
+            'config': {
+                'type': CacheTypes(cache_type).name.lower(),
+            }
+        }
+    else:
+        print('\nNo cache set.')
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:156]: when this else statement is called?

File: core_lib_generator/core_lib_template_generator.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+
+
+def generate_db_entity_template() -> dict:
+    entities = {}
+    add_entity = True
+    while add_entity:
+        is_soft_delete = False
+        is_soft_delete_token = False
+        entity_name = input_str('Enter the name of the entity you\'d like to create: ', 'User')
+        while entity_name in entities:
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:166]: no duplicate logic

File: core_lib_generator/core_lib_template_generator.py [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+                column_default = input_str(f'Enter the default value of column #{i + 1}', ' ')
+
+                columns[column_name] = {
+                    'name': column_name,
+                    'type': DBDatatypes(column_type).name,
+                    'default': column_default
+                }
+
+            is_soft_delete = input_yes_no('Do you want to implement Soft Delete?', False)
+            if is_soft_delete:
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:185]: no duplicate logic

**Review by shay-te — CHANGES_REQUESTED** [review]

File: core_lib_generator/core_lib_template_generator.py (hunk 15 above)
Comment (shubham-shay-te) [reply core_lib_generator/core_lib_template_generator.py:156]: When the cache is selected as empty

File: core_lib_generator/core_lib_template_generator.py (hunk 16 above)
Comment (shubham-shay-te) [reply core_lib_generator/core_lib_template_generator.py:166]: if the user is reentering a name that is already being created it is for that

File: core_lib_generator/core_lib_template_generator.py (hunk 17 above)
Comment (shubham-shay-te) [reply core_lib_generator/core_lib_template_generator.py:185]: this is for asking if `Soft delete` is not selected by user then ` we should not ask for `Soft delete token` right?

File: core_lib_generator/core_lib_template_generator.py (hunk 13 above)
Comment (shubham-shay-te) [reply core_lib_generator/core_lib_template_generator.py:125]: I had seen it in one of your configs so included it, this is for the `core` name actually

File: core_lib/helpers/shell_utils.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def _check_yes_no(value):
+    if value.lower() in ['yes', 'y']:
+        return True
+    elif value.lower() in ['no', 'n']:
+        return False
+
+
+def input_yes_no(title: str, default_value: bool = None) -> bool:
```
[the remark this answers is not in the document]
Comment (shubham-shay-te) [reply core_lib/helpers/shell_utils.py:34]: tried it, but it is failing for some reason, I have another implementation here
```
    result = None
    while result is None:
        user_input = input(f'{title} (yes/no) {_print_default(converted_default_value)}: ')
        user_input = _get_value(user_input, converted_default_value)
        if user_input in ['yes', 'y', 'no', 'n']:
            result = _check_yes_no(user_input)
    return result
```
will this do?

File: core_lib/helpers/shell_utils.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    take_input = True
+    while take_input:
+        user_input = input(f'{title}{_print_default(default_value)}: ')
+        user_input = _check_default(user_input, default_value)
+        if is_int(user_input):
+            return int(user_input)
+
+
+def input_enum(enum_class: enum, title: str, default_value: int = None) -> int:
+    enum_values = set()
```
[the remark this answers is not in the document]
Comment (shubham-shay-te) [reply core_lib/helpers/shell_utils.py:63]: thankyou

File: core_lib_generator/core_lib_template_generator.py [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+        db_password: str,
+        db_port: int,
+        db_host: str = 'localhost',
+        db_log_queries: bool = False,
+        db_create: bool = True,
+        db_pool_recycle: int = 3200,
+        db_pool_pre_ping: bool = False,
+) -> dict:
+    env = {}
+    if not db_host:
```
Comment (shay-te) [inline core_lib_generator/core_lib_template_generator.py:63]: ask is sqlite instead

Comment (shubham-shay-te) [reply core_lib_generator/core_lib_template_generator.py:63]: added the changes to other branch  #59

**Pull request closed without merging**